### PR TITLE
Late arriving data

### PR DESCRIPTION
**Pull request by Anaisdg**
Added a late arriving data guide. 
Thank you!

- [x] Signed the [InfluxData CLA](https://www.influxdata.com/legal/cla/)
  ([if necessary](https://github.com/influxdata/docs-v2/blob/master/CONTRIBUTING.md#sign-the-influxdata-cla))
- [x] Rebased/mergeable

### CODE REVIEW

File: content/resources/how-to-guides/late-arriving-data.md [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+![late arriving data architecture](images/late-arriving-data.png "image_tooltip")
+
+
+The water_level_checksum.flux is a task. It runs every minute. It counts the number of points that exist in the water_level_raw bucket (new count) and compares that count against the count in the water_level_checksum bucket (old count). If the new count from the water_level_raw bucket isn’t equal to the count from the water_level_checksum bucket then an invokable script is called. Specifically, the water_level_process.flux script is called. This script is responsible for recalculating the old count and aggregation.   
+
+
+## Flux scripts in detail: water_level_process.flux
+
+The water_level_process.flux script is an invokable script. It’s responsible for two things:
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:42]:
```suggestion
## Flux scripts in detail

- [water_level_process.flux](#water_level_processflux)
- [water_level_checksum.flux](#water_level_checksumflux)

### water_level_process.flux

`water_level_process.flux` is an invokable script that does two things:
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    |> count()
+    |> to(bucket: "water_level_checksum", timeColumn: "_stop")
+    |> yield(name: "checksums")
+```
+
+
+
+## Flux scripts in detail: water_level_checsum.flux
+
+The water_level_process.flux script is a task. This task is responsible for:
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:72]:
```suggestion
## water_level_checsum.flux

`water_level_process.flux` is a task that does the following:
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+The water_level_checksum.flux is a task. It runs every minute. It counts the number of points that exist in the water_level_raw bucket (new count) and compares that count against the count in the water_level_checksum bucket (old count). If the new count from the water_level_raw bucket isn’t equal to the count from the water_level_checksum bucket then an invokable script is called. Specifically, the water_level_process.flux script is called. This script is responsible for recalculating the old count and aggregation.   
+
+
+## Flux scripts in detail: water_level_process.flux
+
+The water_level_process.flux script is an invokable script. It’s responsible for two things:
+
+
+1. Computing the mean value over a parameterized start and stop times. This mean value is written to the water_level_mean bucket. 
+2. Calculating the count or total numper of points over parameterized start and stop times. This count is written to the water_level_checksum bucket.
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:46]:
```suggestion
1. Computes the mean of values in the time range defined by the `start` and `stop` script parameters and writes the computed mean to the **water_level_mean** bucket. 
2. Computes the count or total number of points in the time range defined by the `start` and `stop` script parameters and writes the count to the **water_level_checksum** bucket. 
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+// Compute and store new checksum for this window
+from(bucket: "water_level_raw")
+    |> range(start: params.start, stop: params.stop)
+    |> group(columns: ["_measurement", "_field", "_stop"])
+    |> count()
+    |> to(bucket: "water_level_checksum", timeColumn: "_stop")
+    |> yield(name: "checksums")
+```
+
+
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:68]:
~~~suggestion
```js
// Compute and store the mean for the window
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> mean()
    |> to(bucket: "water_level_mean", timeColumn: "_stop")
    |> yield(name: "means")

// Compute and store the new checksum for this window
from(bucket: "water_level_raw")
    |> range(start: params.start, stop: params.stop)
    |> group(columns: ["_measurement", "_field", "_stop"])
    |> count()
    |> to(bucket: "water_level_checksum", timeColumn: "_stop")
    |> yield(name: "checksums")
```
~~~

File: content/resources/how-to-guides/late-arriving-data.md [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+The water_level_process.flux script is a task. This task is responsible for:
+
+
+1. Counting the number of points in the water_level_raw bucket (new count) for the last hour across 1minute windows.
+2. Invoking the water_level_process.flux invokable script to calculate a new mean and a new count across 1 minute windows. 
+3. Gathering the previous count in the water_level_checksum bucket (old count) for the last hour. 
+4. Comparing an old count against a new count with a join.  
+5. Filtering for where the counts do not match.
+6. Invoking the water_level_process.flux invokable script to recompute the mean and count for every 1 min window where the counts don't match.
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:80]:
```suggestion
1. Counts the number of points in the **water_level_raw** bucket (new count) for the last hour across one minute windows.
2. Invokes the `water_level_process.flux` invokable script to calculate a new mean and a new count across one minute windows. 
3. Gathers the previous count in the **water_level_checksum** bucket (old count) for the last hour. 
4. Joins the old and new streams and compares the old count against a new count.  
5. Filters for counts that do not match.
6. Invokes the `water_level_process.flux` invokable script to recompute the mean and count for every one minute window where the counts do not match. 
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 6]
```diff
@@ -0,0 +1,178 @@
...
+option task = {name: "water_level_checksum", every: 1m, offset: 10s}
+```
+
+
+This `every` parameter also determines the window period for which the mean and counts will be computed over. 
+
+The invokeScripts() function is a custom function that invokes the water_level_process.flux invokable script. \
+`invokeScript = (start, stop) =>`
+
+
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:95]:
```suggestion
#### Task details

- The `task` option provides configuration settings for the task:
    - `name`: Provides a name for the task.
    - `every`: Defines how often the task runs (every one minute) and, in this case, the window interval used to compute means and counts.
    - `offset`: Defines how much time to wait before executing the task. _The offset does not change the time range queried by the task. _
- `invokeScripts()` is a custom function that invokes the `water_level_process.flux` invokable script.
    - `start` and `stop` parameters are required.
    - `scriptID` is required.
    - Store your InfluxDB API token as an InfluxDB secret and use the `secrets` package to retrieve the token.
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+```
+    requests.post(
+        // The script ID here has been hardcoded here
+        url: "https://eastus-1.azure.cloud2.influxdata.com/api/v2/scripts/095fabd404108000/invoke",
+
+        headers: ["Authorization": "Token ${token}", "Accept": "application/json", "Content-Type": "application/json"],
+        body: json.encode(v: {params: {start: string(v: start), stop: string(v: stop)}}),
+    )
+```
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:104]: I'd parameterize the script ID.

~~~suggestion
```
option task = {name: "water_level_checksum", every: 1m, offset: 10s}

invokeScript = (start, stop, scriptID) =>
    requests.post(
        url: "https://cloud2.influxdata.com/api/v2/scripts/${scriptID}/invoke",

        headers: ["Authorization": "Token ${token}", "Accept": "application/json", "Content-Type": "application/json"],
        body: json.encode(v: {params: {start: string(v: start), stop: string(v: stop)}}),
    )
```
~~~

We also need to instruct them where to find the script ID.

File: content/resources/how-to-guides/late-arriving-data.md [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+
+Remember this invokable script is responsible for both calculating the mean and count and writing it to the water_level_mean and water_level_checksum buckets respectively.  You must pass in values for the start and stop parameters. The token is imported with the secrets manager.
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:107]:
```suggestion
```

File: content/resources/how-to-guides/late-arriving-data.md [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    |> filter(fn: (r) => r.old_count != r.new_count)
+    |> map(
+        fn: (r) => {
+            response = invokeScript(start: date.sub(d: every, from: r._time), stop: r._time)
+
+            return {r with code: response.statusCode}
+        },
+    )
+    |> yield(name: "diffs")
+```
```
Comment (sanderson) [inline content/resources/how-to-guides/late-arriving-data.md:178]: Explain the task in the proposed "Task details" section, but show the task in it's entirety.

**Review by sanderson — CHANGES_REQUESTED** [review]
Some notes:

- Your image needs to me moved to `/img/resources/`.
- Flux scripts are explained in detail, but there's no process outline:
    - How do I add the task?
    - How do I add the invokable script?
    - Where do I get the invokable script ID?
    - How do I create a bucket?
    - etc.

File: content/resources/how-to-guides/late-arriving-data.md (hunk 9 above)
Comment (Anaisdg) [reply content/resources/how-to-guides/late-arriving-data.md]: Thank you! All comments have been addressed.

**Comment by sanderson** [comment]
@Anaisdg just one last change that needs to be made here. All the images need to be stored in the `/static/img/` directory, so you need move `content/resources/how-to-guides/images/late-arriving-data.png` to `static/img/resources/late-arriving-data.png`

**Comment by Anaisdg** [comment]
Thank you @sanderson

**Review by sanderson — APPROVED** [review]

**Pull request merged**